### ISSUE TITLE
Babel analyzer running before `each()` from `@embroider/macros` gets a chance to transpose imports

### ISSUE THREAD

**Issue by Panman82** [ef4/ember-auto-import#477]
In discussion over on #475, it seems that (undocumented) `each()` from `@embroider/macros` should [transpose imports when analyzable at build-time](https://github.com/embroider-build/embroider/blob/v0.48.1/packages/macros/src/babel/each.ts#L56-L63). Ex:

```js
import { each, importSync } from '@embroider/macros';

// FROM
for (const asset of each(['package-name/something.js','package-name/something-else.js'])) {
  importSync(asset);
}

// TO
importSync('package-name/something.js');
importSync('package-name/something-else.js');
```

However, the [analyzer in ember-auto-import](https://github.com/ef4/ember-auto-import/blob/v2.2.4/packages/ember-auto-import/ts/analyzer-plugin.ts#L52-L58) enforces the [dynamic `import()` requirements](https://github.com/emberjs/rfcs/blob/73685c28378118bebb5e359b80e00b839a99f622/text/0507-embroider-v2-package-format.md#supported-subset-of-dynamic-import-syntax) for `importSync()` before `each()` gets a chance to transpose those imports. Mainly, the above example results in a [string literal error from the analyzer](https://github.com/ef4/ember-auto-import/blob/v2.2.4/packages/ember-auto-import/ts/analyzer-plugin.ts#L111). Ex:

```cli
Build Error (broccoli-persistent-filter:Babel > [Babel: addon-name]) in addon-name/components/foobar.js

/path/to/addon-name/components/foobar.js: import() is only allowed to contain string literals or template string literals
  1 | import { each, importSync } from '@embroider/macros';
  2 |
  3 | for (const asset of each(['package-name/something.js','package-name/something-else.js'])) {
> 4 |   importSync(asset);
    |   ^^^^^^^^^^^^^^^^^
  5 | }
```

**Comment by Panman82** [comment]
So I was going to try tackling this but don't really know where to go @ef4 . From my prior digging, it seems like I need to avoid the dynamic import requirements check [here](https://github.com/ef4/ember-auto-import/blob/v2.2.4/packages/ember-auto-import/ts/analyzer-plugin.ts#L56) by using the same condition that `each()` would use [here](https://github.com/embroider-build/embroider/blob/v0.48.1/packages/macros/src/babel/each.ts#L48). However, I'm unsure if `state.opts.mode` is available in EAI. Ex:

```js
if (state.opts.mode !== 'compile-time') {
  state.imports.push(processImportCallExpression(path, false));
}
state.handled.add(path.node);
```

**Comment by ef4** [comment]
Here's what I think is going on.

The `each` macro gets handled as we enter the "for of" statement:

https://github.com/embroider-build/embroider/blob/6a2e220f08db8558af8366e9d186d2f5df86953f/packages/macros/src/babel/macros-babel-plugin.ts#L49

The analyzer detects `importSync` as we enter a `CallExpression`:

https://github.com/ef4/ember-auto-import/blob/1d24f61a3a6f468689e488c5c8e657ad9e22247f/packages/ember-auto-import/ts/analyzer-plugin.ts#L62

Given this ordering, the `each` should already happen first, since you would enter the `for of` before entering its body.

But the `each` doesn't always get unrolled. It depends on whether the macros are in run-time vs compile-time mode:

https://github.com/embroider-build/embroider/blob/6a2e220f08db8558af8366e9d186d2f5df86953f/packages/macros/src/babel/each.ts#L52

You can probably confirm this by testing a production build. Your example might work in prod but not dev.

In development we use run-time mode, which was motivated by the desire to have a single build that can give different answers for macros like `isTesting()` and `getGlobalConfig().fastboot?.isRunning` depending on the context you're running the code in. 

In production we always use build-time mode, and do separate optimized builds for each possible context. The rationale is that this is slower but produces optimal bundles.

The issue here is that run-time vs built-time is too global. Only a few macros (like `isTesting()`) should support having run-time implementations, and others like `each` and `macroCondition` should react to their argument. For example, if `each` sees an argument that is a macro with a build-time implementation, it should always use its own build-time implementation and unroll the loop. Whereas if it sees an argument that is a macro with a run-time implementation, it should use its own run-time implementation.

So in summary, we have a refactor needed in embroider macros that would extend the protocol macros use for understanding each other's values. The `EvaluateResult` type would get extended so that in addition to distinguishing between "confident-with-a-value" and "not confident", it could do "confident but with a value that's being deferred until runtime". It's probably not a huge amount of code to change, but it touches pretty deep parts.

(This would fix another open issue too, which is that you can't currently use `if (macroCondition(dependencySatisfies(...))){  ... }` to guard invalid code in development, since it currently gets a run-time implementation that always leaves the code in place, even though `dependencySatisfies` can't really ever change at runtime.)

**Comment by Panman82** [comment]
> But the each doesn't always get unrolled. It depends on whether the macros are in run-time vs compile-time mode:

Ah, I was thinking that bit referenced if the array was run-time or compile-time accessible.

> You can probably confirm this by testing a production build. Your example might work in prod but not dev.

Yep, works without error when building for production. `ember server -prod`

> It's probably not a huge amount of code to change, but it touches pretty deep parts.

Yeah, I'm having a tough time groking how all this fits together. And to me, I'm not sure why `each()` would ever [_not_ get unrolled](https://github.com/embroider-build/embroider/blob/6a2e220f08db8558af8366e9d186d2f5df86953f/packages/macros/src/babel/each.ts#L52-L54).

**Comment by Panman82** [comment]
Closing this ticket since the root issue is actually `@embroider/macro` based and I opened up a new ticket over on embroider: https://github.com/embroider-build/embroider/issues/1063
